sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCoercion.scala

```diff
@@ -357,6 +357,9 @@ object CollationTypeCoercion {
     case collate: Collate =>
       Some(addContextToStringType(collate.dataType, Explicit))
 
+    case expr @ (_: NamedExpression | _: SubqueryExpression | _: VariableReference) =>
```

Comment:
is there any reason to move this `case` up?

Reply:
Yes, I'd like to have these sorted by collation strength, so that implicit case comes before the cast. I didn't think about it when I added the `cast` case...